src/tasks/AotCompilerTask/MonoAOTCompiler.cs

```diff
@@ -605,6 +627,27 @@ private PrecompileArguments GetPrecompileArgumentsFor(ITaskItem assemblyItem, st
             aotArgs.Add($"static");
         }
 
+        if (UseDirectPInvoke)
+        {
+            aotArgs.Add($"direct-pinvoke");
+        }
+
+        if (DirectPInvokes.Length > 0)
+        {
+            foreach (var directPInvoke in DirectPInvokes)
+            {
+                aotArgs.Add($"direct-pinvokes={directPInvoke}");
```

Comment:
The name of the argument makes me thing that you could add multiple symbols to the same argument, but looks like that is not the case, each argument only accepts one module!entrypoint so if you need multiple you would need to repeat the argument, like direct-pinvokes=kernel32!Sleep, direct-pinvokes=kernel32!MessageBox or is it possible to do all in one, direct-pinvokes=kernel32!Sleep;kernel32!MessageBox?

If not, then maybe the argument should not indicate that it handles multiple entries, so should probably be named differently. Since direct-pinvoke is already reserved, maybe direct-pinvoke-name= could be an alternative?

The argument to the task could still be named DirectPInvokes since it will be mapped to an item group in msbuild <DirectPInvoke>.

Reply:
Good point, this slipped past me. I think it might be more straightforward to have the all-in-one parameter option `direct-pinvokes=kernel32!Sleep;kernel32!MessageBox` if one were to call the mono aot compiler directly. To implement such, I replaced the corresponding handle `opts->direct_pinvokes = g_list_append (opts->direct_pinvokes, g_strdup (arg + strlen ("direct-pinvokes=")));` in `mono_aot_parse_options` with 
```
char *direct_pinvokes = g_strdup (arg + strlen ("direct-pinvokes="));
gchar *direct_pinvoke_ctx = NULL;
gchar *direct_pinvoke = strtok_r (direct_pinvokes, ",", direct_pinvoke_ctx);
while (direct_pinvoke) {
	opts->direct_pinvokes = g_list_append (opts->direct_pinvoke, g_strdup (direct_pinvoke));
	direct_pinvoke = strtok_r (NULL, ",", direct_pinvoke_ctx);
}
g_free (direct_pinvokes);
```
To process every direct_pinvoke token in the provided argument. One question I have is how does the data stored in a `MonoAotCompile` instance's `MonoAotOptions` field's fields get freed/cleared? Because I allocated mem `g_strdup (direct_pinvoke)` to store in the list, I would need to ensure that gets freed at the end of its scope of use. Should I add something to `acfg_free`?

Reply:
I extended the mono aot compiler to accept the parameter as both `direct-pinvokes=kernel32!Sleep, direct-pinvokes=kernel32!MessageBox` and `direct-pinvokes=kernel32!Sleep;kernel32!MessageBox`. Changed the respective MonoAOTCompiler Task's argument processing to pass as the latter `aotArgs.Add($"direct-pinvokes={string.Join(",", DirectPInvokes)}");`

Reply:
See comment above, looks like you used , as delimitator between multiple direct-pinvokes instead of ; ?

Reply:
> To process every direct_pinvoke token in the provided argument. One question I have is how does the data stored in a `MonoAotCompile` instance's `MonoAotOptions` field's fields get freed/cleared? Because I allocated mem `g_strdup (direct_pinvoke)` to store in the list, I would need to ensure that gets freed at the end of its scope of use. Should I add something to `acfg_free`?

Currently none of the aot_opts gets freed in acfg_free, so its currently a per compiled assembly leak, but since aot compiler normally is run with 1 assembly in each invocation it has probably not been an issue, but maybe it could be wise to add a function call in acfg_free that frees all aot options.

Reply:
I added `aot_opts_free` which is called near the end of `acfg_free` that will free all fields that are allocated via `g_strdup` or `g_list_append`. I noticed that MonoAotOptions is first initialized https://github.com/dotnet/runtime/blob/210a7a9feec64b7fac5147656cddb199ec90cf75/src/mono/mono/mini/aot-compiler.c#L14392 (after the MonoAotCompile instance is initially allocated with `g_new0`, which I'm guessing just allocates a pointer for `aot_opts` and not the actual space for MonoAotOptions`?) Does this mean it overwrites memory its not assigned or was enough memory allocated for `MonoAotOptions` when using `g_new0` on `MonoAotCompile`? I noticed a problem when trying to `g_free (aot_opts)` at the end of `aot_opts_free`

Reply:
You shouldn't free acfg-aot_opts since that memory is a struct part of struct MonoAotCompile so the complete memory for MonoAotCompile (including MonoAotOptions) gets allocated and freed in acfg_create and acfg_free. Just free the actually pointers to allocated memory that MonoAotOptions fields points to.